certgen/certgen.go

```diff
@@ -0,0 +1,171 @@
+// certificate chains using its RootCert and RootKey.
+type CA struct {
+	RootCert   *x509.Certificate
+	RootKey    crypto.Signer
+	CertConfig CertificateConfig
+}
+
+// IssueCertificate creates a new leaf certificate, issued by the root specified
+// in the RootCert and RootKey fields of the CA, and configured using the
+// CertConfig in the CA.
+func (ca *CA) IssueCertificate() (*x509.Certificate, error) {
+	// TODO(katjoyce): Make the type of key generated configurable.
+	key, err := rsa.GenerateKey(crand.Reader, keySize)
+	if err != nil {
+		return nil, fmt.Errorf("error generating key pair: %s", err)
+	}
+
+	template, err := leafTemplate(ca.CertConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf template: %s", err)
+	}
+
+	leafDER, err := x509.CreateCertificate(crand.Reader, template, ca.RootCert, key.Public(), ca.RootKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf certificate: %s", err)
+	}
+
+	leaf, err := x509.ParseCertificate(leafDER)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing leaf certificate DER: %s", err)
+	}
+
+	return leaf, nil
+}
+
+func leafTemplate(c CertificateConfig) (*x509.Certificate, error) {
+	sn, err := randSerialNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	notAfter := randNotAfter(c.NotAfterInterval)
+
+	return &x509.Certificate{
+		SerialNumber: sn,
+		Subject: pkix.Name{
+			Country:            []string{c.SubjectCountry},
+			Organization:       []string{c.SubjectOrganization},
+			OrganizationalUnit: []string{c.SubjectOrganizationalUnit},
+			Locality:           []string{c.SubjectLocality},
+			CommonName:         c.SubjectCommonName,
+		},
+		NotBefore:          notAfter.Add(-certValidity),
+		NotAfter:           notAfter,
+		SignatureAlgorithm: c.SignatureAlgorithm,
+
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+		DNSNames:              []string{c.SubjectCommonName, extendedDNSSAN(c.Prefix, c.SubjectCommonName)},
+	}, nil
+}
+
+func randSerialNumber() (*big.Int, error) {
+	i := big.NewInt(0)
+	return crand.Int(crand.Reader, i.SetUint64(math.MaxUint64))
+}
+
+func randNotAfter(notAfterInterval *interval.Interval) time.Time {
+	if notAfterInterval == nil {
+		// TODO(katjoyce): Make this random.
+		return timeNowUTC().Add(certValidity)
+	}
+
+	rand.Seed(timeNowUTC().UnixNano())
+
+	start := notAfterInterval.Start.Unix()
+	end := notAfterInterval.End.Unix()
+	delta := end - start
+
+	return time.Unix(start+rand.Int63n(delta), 0)
+}
+
+// extendedDNSSAN creates a string to be used in the DNSNames SAN.  The string
+// created is or the format <hour>.<day>.<month>.<year>.<prefix>.<url> where the
+// time elements are based on the time now.  For example, if
+// extendedDNSSAN(abc, xyz) was called at 2019-03-25 12:00 UTC, it would return
+// 12.25.march.2019.abc.xyz
```

Comment:
It'd be nicer to use a 0-prefixed numerical month than a bit of text in the middle of what is otherwise an easily sortable string!

your example: 12.25.03.2019.abc.xyz

(btw abc.xyz is a real Google-owned domain, I'd perhaps err on the side of caution and use something different as your example)

Reply:
Ahaha, good to know!  Changed the example, and made the date fully numerical.